File: dryad-core/src/main/java/io/growing/dryad/annotation/Configuration.java
package io.growing.dryad.annotation;

import io.growing.dryad.parser.ConfigParser;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Component:
 * Description:
 * Date: 16/3/18
 *
 * @author Andy Ai
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface Configuration {
    String name();

    Class<? extends ConfigParser<?>> parser();
}

File: dryad-core/src/main/scala/io/growing/dryad/internal/Configuration.scala
package io.growing.dryad.internal

import java.util.concurrent.atomic.AtomicReference

/**
 * Component:
 * Description:
 * Date: 16/3/22
 *
 * @author Andy Ai
 */
trait Configuration {
  def name: String

  def payload: String

  def version: Long

  def namespace: String

  def group: String

}

final case class ConfigurationValue(name: String, payload: String, version: Long, namespace: String, group: String) extends Configuration

class ConfigurationReference(underlying: Configuration) extends Configuration {
  private val ref = new AtomicReference(underlying)

  override def name: String = ref.get.name

  override def payload: String = ref.get.payload

  override def version: Long = ref.get.version

  override def namespace: String = ref.get.namespace

  override def group: String = ref.get.group

  def set(underlying: Configuration): Unit = {
    ref.set(underlying)
  }
}
